File: inference-v3/src/magnitude_engine/artifacts/mlx.py
"""MLX affine Safetensors artifacts. No MLX execution dependency."""

import hashlib
import json
import math
import struct
from contextlib import ExitStack
from dataclasses import dataclass
from pathlib import Path

from pydantic import TypeAdapter

from magnitude_engine.artifacts.identity import ArtifactIdentity
from magnitude_engine.artifacts.weights import WeightDescriptor
from magnitude_engine.platform.execution import DType, TensorSpec
from magnitude_engine.platform.storage import FileSource


@dataclass(frozen=True)
class StoredTensor:
    source: FileSource
    offset: int
    spec: TensorSpec

    def read(self) -> bytes:
        return self.source.read(self.offset, self.spec.nbytes)


class MLXArtifact:
    """Own an immutable directory snapshot and validate encoded tensor ranges.

    Artifact interpretation is host work; all numerical conversion and execution
    uses the selected TileLang device. Floating tensors retain their stored dtype.
    """

    def __init__(self, path: str):
        self.path = Path(path).expanduser()
        with ExitStack() as cleanup:
            config = FileSource(self.path / "config.json")
            cleanup.callback(config.close)
            content = config.read(0, config.size)
            self.config = TypeAdapter(dict[str, object]).validate_json(content)
            quantization = self.config.get("quantization")
            if quantization != {"group_size": 64, "bits": 4, "mode": "affine"}:
                raise ValueError("MLX artifact requires uniform affine Q4 group-64 weights")
            digest = hashlib.sha256(content)
            self.tensors: dict[str, StoredTensor] = {}
            for tensor_path in sorted(self.path.glob("*.safetensors")):
                source = FileSource(tensor_path)
                cleanup.callback(source.close)
                digest.update(tensor_path.name.encode() + b"\x00" + source.digest().encode())
                header_size = struct.unpack("<Q", source.read(0, 8))[0]
                if header_size > 64 * 1024**2 or 8 + header_size > source.size:
                    raise ValueError("invalid Safetensors header extent")
                header = json.loads(source.read(8, header_size))
                ranges = []
                for name, entry in header.items():
                    if name == "__metadata__":
                        continue
                    if name in self.tensors:
                        raise ValueError(f"duplicate Safetensors tensor {name}")
                    dtype = {"U32": DType.U32, "BF16": DType.BF16, "F32": DType.F32}.get(
                        entry["dtype"]
                    )
                    if dtype is None:
                        raise ValueError(
                            f"unsupported Safetensors dtype for {name}: {entry['dtype']}"
                        )
                    shape = tuple(entry["shape"])
                    if any(type(n) is not int or n <= 0 for n in shape):
                        raise ValueError(f"invalid Safetensors shape for {name}")
                    spec = TensorSpec(shape, dtype)
                    start, end = entry["data_offsets"]
                    if (
                        type(start) is not int
                        or type(end) is not int
                        or start < 0
                        or end - start != spec.nbytes
                        or end > source.size - 8 - header_size
                    ):
                        raise ValueError(f"invalid Safetensors range for {name}")
                    ranges.append((start, end))
                    self.tensors[name] = StoredTensor(source, 8 + header_size + start, spec)
                ordered = sorted(ranges)
                if any(b[0] < a[1] for a, b in zip(ordered, ordered[1:], strict=False)):
                    raise ValueError("overlapping Safetensors ranges")
            if not self.tensors:
                raise ValueError("MLX artifact has no tensors")
            self.identity = ArtifactIdentity(digest.hexdigest())
            self._cleanup = cleanup.pop_all()

    def descriptor(self, name: str, shape: tuple[int, ...]) -> WeightDescriptor:
        stored = self.tensors[name]
        if stored.spec.dtype == DType.U32:
            if len(shape) != 2 or shape[1] % 64:
                raise ValueError(f"invalid affine geometry: {name}")
            n, k = shape
            if stored.spec.shape != (n, k // 8) or not name.endswith(".weight"):
                raise ValueError(f"affine code shape differs: {name}")
            for suffix in ("scales", "biases"):
                if self.tensors[name.removesuffix("weight") + suffix].spec != TensorSpec(
                    (n, k // 64), DType.BF16
                ):
                    raise ValueError(f"affine parameter shape/dtype differs: {name}")
        elif math.prod(stored.spec.shape) != math.prod(shape):
            raise ValueError(f"floating parameter shape differs: {name}")
        return WeightDescriptor(name=name, shape=shape)

    def close(self) -> None:
        self._cleanup.close()

File: inference-v4/engine/src/weights/mlx.rs
//! Device-free interpretation of V3's uniform affine Q4 group-64 artifacts.
use super::{
    descriptor::{ArtifactIdentity, Stored, StoredTensor, Transform, WeightDescriptor},
    safetensors,
    source::FileSource,
    Error,
};
use seismic::DType;
use sha2::{Digest, Sha256};
use std::{collections::BTreeMap, path::Path, sync::Arc};

pub struct MlxArtifact {
    config: serde_json::Value,
    identity: ArtifactIdentity,
    tensors: BTreeMap<String, StoredTensor>,
}
fn invalid(s: impl Into<String>) -> Error {
    Error::Invalid(s.into())
}
impl MlxArtifact {
    pub fn config(&self) -> &serde_json::Value {
        &self.config
    }
    pub fn identity(&self) -> ArtifactIdentity {
        self.identity
    }
    pub fn open(path: impl AsRef<Path>) -> Result<Self, Error> {
        let path = path.as_ref();
        let config_source = FileSource::open(path.join("config.json"))?;
        let config_bytes = config_source.read(
            0,
            usize::try_from(config_source.size())
                .map_err(|_| invalid("configuration exceeds host address range"))?,
        )?;
        let config: serde_json::Value = serde_json::from_slice(&config_bytes)
            .map_err(|e| invalid(format!("invalid MLX configuration: {e}")))?;
        if config.get("quantization")
            != Some(&serde_json::json!({"group_size":64,"bits":4,"mode":"affine"}))
        {
            return Err(invalid(
                "MLX artifact requires uniform affine Q4 group-64 weights",
            ));
        }
        let mut paths = std::fs::read_dir(path)?
            .map(|e| e.map(|e| e.path()))
            .collect::<Result<Vec<_>, _>>()?;
        paths.retain(|p| p.extension().is_some_and(|e| e == "safetensors"));
        paths.sort();
        let mut digest = Sha256::new();
        digest.update(&config_bytes);
        let mut tensors = BTreeMap::new();
        for path in paths {
            let name = path
                .file_name()
                .and_then(|n| n.to_str())
                .ok_or_else(|| invalid("non-UTF8 artifact shard name"))?;
            let source = Arc::new(FileSource::open(&path)?);
            let directory = safetensors::read_directory(&mut source.reader())?;
            digest.update(name.as_bytes());
            digest.update([0]);
            digest.update(ArtifactIdentity(source.digest()?).to_string().as_bytes());
            for t in directory.tensors {
                let name = t.name;
                let stored = StoredTensor {
                    source: source.clone(),
                    offset: t.offset,
                    nbytes: t.nbytes,
                    dtype: t.dtype,
                    shape: t.shape,
                };
                if tensors.insert(name.clone(), stored).is_some() {
                    return Err(invalid(format!("duplicate Safetensors tensor {name:?}")));
                }
            }
        }
        if tensors.is_empty() {
            return Err(invalid("MLX artifact has no tensors"));
        }
        Ok(Self {
            config,
            identity: ArtifactIdentity(digest.finalize().into()),
            tensors,
        })
    }
    pub fn tensors(&self) -> &BTreeMap<String, StoredTensor> {
        &self.tensors
    }
    pub fn descriptor(&self, name: &str, shape: &[u64]) -> Result<WeightDescriptor, Error> {
        self.validate(name, shape)?;
        Ok(WeightDescriptor {
            name: name.into(),
            shape: shape.into(),
            transform: Transform::Identity,
        })
    }
    fn tensor(&self, name: &str) -> Result<&StoredTensor, Error> {
        self.tensors
            .get(name)
            .ok_or_else(|| invalid(format!("missing tensor {name:?}")))
    }
    fn validate(&self, name: &str, shape: &[u64]) -> Result<(), Error> {
        if shape.contains(&0) {
            return Err(invalid("weight role requires positive dimensions"));
        }
        let elements = shape
            .iter()
            .try_fold(1u64, |a, n| a.checked_mul(*n))
            .ok_or_else(|| invalid("weight role shape overflows"))?;
        let entry = self.tensor(name)?;
        if entry.dtype == DType::U32 {
            let [n, k] = shape else {
                return Err(invalid(format!("invalid affine geometry {name:?}")));
            };
            if k % 64 != 0 || entry.shape != [*n, k / 8] || !name.ends_with(".weight") {
                return Err(invalid(format!("affine code shape differs {name:?}")));
            }
            for suffix in ["scales", "biases"] {
                let coefficient =
                    self.tensor(&format!("{}{suffix}", name.strip_suffix("weight").unwrap()))?;
                if coefficient.dtype != DType::BF16 || coefficient.shape != [*n, k / 64] {
                    return Err(invalid(format!(
                        "affine coefficient shape/dtype differs {name:?}"
                    )));
                }
            }
        } else if elements != entry.nbytes / u64::from(entry.dtype.bytes()) {
            return Err(invalid(format!(
                "floating parameter shape differs {name:?}"
            )));
        }
        Ok(())
    }
    pub fn stored(&self, descriptor: &WeightDescriptor) -> Result<Stored, Error> {
        self.validate(&descriptor.name, &descriptor.shape)?;
        let entry = self.tensor(&descriptor.name)?;
        if entry.dtype != DType::U32 {
            return Ok(Stored::Dense(entry.clone()));
        }
        let prefix = descriptor.name.strip_suffix("weight").unwrap();
        Ok(Stored::AffinePlanes {
            bits: 4,
            group: 64,
            codes: entry.clone(),
            scales: self.tensor(&format!("{prefix}scales"))?.clone(),
            biases: self.tensor(&format!("{prefix}biases"))?.clone(),
        })
    }
}
